File: plugins/guardian/ruleset.py
"""Graph-synced rule cache.

The :class:`Ruleset` is built entirely from DKG query results, merged from two
tiers with strict precedence:

* ``verifiable-memory`` (the curated public threat graph) → rules tagged
  ``source: "public"``. The source of truth: matches are CONFIRMED and, in
  block mode, blockable.
* ``shared-working-memory`` (the community pool) → rules tagged
  ``source: "community"``. Checked only when the public graph doesn't already
  cover the identifier: matches are flagged but NEVER block — anyone can write
  to the community pool, so it must not be able to stop tool calls.

It is cached to ``$GUARDIAN_HOME/ruleset.json`` and refreshed lazily:
:func:`get` returns the cached ruleset immediately and, if the cache is older
than ``sync_interval``, kicks off a single non-blocking background refresh
(guarded by a file lock so only one refresher runs across processes). Every
path fails open to the last-good (or empty) ruleset.
"""

from __future__ import annotations

import json
import logging
import re
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

from . import constants
from .config import GuardianConfig, load_guardian_config
from .dkg_client import DkgClient, extract_binding

logger = logging.getLogger(__name__)

# SPARQL that pulls every threat's queryable fields in one shot.
_THREATS_SPARQL = """
PREFIX g: <http://umanitek.ai/ontology/guardian/>
PREFIX schema: <http://schema.org/>
SELECT ?identifier ?severity ?name ?pattern ?toolName ?argShape
       ?packageName ?packageVersion ?packageEcosystem ?advisoryId ?curated
       ?category ?skillName ?skillVersion ?dangerShape
WHERE {
  ?threat g:identifier ?identifier .
  OPTIONAL { ?threat g:severity ?severity . }
  OPTIONAL { ?threat schema:name ?name . }
  OPTIONAL { ?threat g:pattern ?pattern . }
  OPTIONAL { ?threat g:toolName ?toolName . }
  OPTIONAL { ?threat g:argShape ?argShape . }
  OPTIONAL { ?threat g:packageName ?packageName . }
  OPTIONAL { ?threat g:packageVersion ?packageVersion . }
  OPTIONAL { ?threat g:packageEcosystem ?packageEcosystem . }
  OPTIONAL { ?threat schema:identifier ?advisoryId . }
  OPTIONAL { ?threat g:curated ?curated . }
  OPTIONAL { ?threat g:category ?category . }
  OPTIONAL { ?threat g:skillName ?skillName . }
  OPTIONAL { ?threat g:skillVersion ?skillVersion . }
  OPTIONAL { ?threat g:dangerShape ?dangerShape . }
}
LIMIT 2000
"""


@dataclass
class Ruleset:
    """Compiled detection rules. See :mod:`detection` for how each is used."""

    injection: List[Dict[str, Any]] = field(default_factory=list)
    escalation: List[Dict[str, Any]] = field(default_factory=list)
    dependency: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    fileaccess: List[Dict[str, Any]] = field(default_factory=list)
    skill: List[Dict[str, Any]] = field(default_factory=list)
    synced_at: float = 0.0

    def counts(self) -> Dict[str, int]:
        return {
            "injection": len(self.injection),
            "escalation": len(self.escalation),
            "dependency": len(self.dependency),
            "fileaccess": len(self.fileaccess),
            "skill": len(self.skill),
        }


# ---------------------------------------------------------------------------
# Build from query bindings
# ---------------------------------------------------------------------------


def _row_to_rule(row: Dict[str, Any], source: str = "public") -> Optional[tuple]:
    """Map one SPARQL binding row to ``(category, key, rule)`` or ``None``.

    *source* tags the rule's trust tier: ``"public"`` (verifiable-memory, the
    curated source of truth) or ``"community"`` (shared-working-memory).
    """
    identifier = extract_binding(row.get("identifier"))
    if not identifier:
        return None
    severity = constants.normalize_severity(extract_binding(row.get("severity")), "high")
    name = extract_binding(row.get("name")) or identifier
    if identifier.startswith("injection:"):
        pattern_src = extract_binding(row.get("pattern"))
        if not pattern_src:
            return None
        try:
            compiled = re.compile(pattern_src, re.IGNORECASE)
        except re.error as exc:
            logger.debug("guardian: skipping bad injection pattern %s: %s", identifier, exc)
            return None
        return ("injection", identifier, {
            "identifier": identifier,
            "pattern": compiled,
            "pattern_src": pattern_src,
            "severity": severity,
            "name": name,
            "source": source,
        })
    if identifier.startswith("escalation:"):
        tool_name = extract_binding(row.get("toolName"))
        arg_shape = extract_binding(row.get("argShape"))
        if not tool_name or not arg_shape:
            return None
        return ("escalation", identifier, {
            "identifier": identifier,
            "toolName": tool_name,
            "argShape": arg_shape,
            "severity": severity,
            "name": name,
            "source": source,
        })
    if identifier.startswith("dep:"):
        eco = extract_binding(row.get("packageEcosystem")).lower()
        pkg = extract_binding(row.get("packageName")).lower()
        ver = extract_binding(row.get("packageVersion"))
        if not (eco and pkg and ver):
            # Fall back to parsing the identifier: dep:{eco}:{name}@{version}
            try:
                _, rest = identifier.split(":", 1)
                eco2, tail = rest.split(":", 1)
                pkg2, ver2 = tail.rsplit("@", 1)
                eco, pkg, ver = eco2.lower(), pkg2.lower(), ver2
            except ValueError:
                return None
        key = f"{eco}:{pkg}@{ver}"
        return ("dependency", key, {
            "identifier": identifier,
            "ecosystem": eco,
            "packageName": pkg,
            "packageVersion": ver,
            "advisoryId": extract_binding(row.get("advisoryId")),
            "severity": severity,
            "name": name,
            "source": source,
        })
    if identifier.startswith("fileaccess:"):
        tool_name = extract_binding(row.get("toolName"))
        category = extract_binding(row.get("category"))
        if not (tool_name and category):
            # Fall back to parsing: fileaccess:{tool}:{category}
            try:
                _, tool_name, category = identifier.split(":", 2)
            except ValueError:
                return None
        return ("fileaccess", identifier, {
            "identifier": identifier,
            "toolName": tool_name.strip().lower(),
            "category": category.strip().lower(),
            "severity": severity,
            "name": name,
            "source": source,
        })
    if identifier.startswith("skill:"):
        rule = {
            "identifier": identifier,
            "skillName": extract_binding(row.get("skillName")),
            "skillVersion": extract_binding(row.get("skillVersion")),
            "dangerShape": extract_binding(row.get("dangerShape")),
            "severity": severity,
            "name": name,
            "source": source,
        }
        return ("skill", identifier, rule)
    return None


def build_from_rows(rows: List[Dict[str, Any]], source: str = "public") -> Ruleset:
    """Build a :class:`Ruleset` from ``(rows, source)`` pairs or plain rows.

    *rows* may be a flat list of binding rows (all tagged *source*) or a list
    of ``(row, source)`` tuples as produced by :func:`refresh`. Precedence is
    identifier-first-wins with public beating community, so a community row can
    never shadow (or escalate/downgrade) a curated public rule.
    """
    rs = Ruleset(synced_at=time.time())
    inj_seen: set = set()
    esc_seen: set = set()
    fa_seen: set = set()
    skill_seen: set = set()
    for item in rows:
        if isinstance(item, tuple):
            row, row_source = item
        else:
            row, row_source = item, source
        mapped = _row_to_rule(row, row_source)
        if not mapped:
            continue
        category, key, rule = mapped
        if category == "injection":
            if key not in inj_seen:
                inj_seen.add(key)
                rs.injection.append(rule)
        elif category == "escalation":
            if key not in esc_seen:
                esc_seen.add(key)
                rs.escalation.append(rule)
        elif category == "fileaccess":
            if key not in fa_seen:
                fa_seen.add(key)
                rs.fileaccess.append(rule)
        elif category == "skill":
            if key not in skill_seen:
                skill_seen.add(key)
                rs.skill.append(rule)
        else:
            existing = rs.dependency.get(key)
            # Public (curated) rules always win over community rows.
            if existing is None or (
                existing.get("source") == "community" and rule.get("source") == "public"
            ):
                rs.dependency[key] = rule
    return rs


# ---------------------------------------------------------------------------
# Cache persistence
# ---------------------------------------------------------------------------


def _cache_path() -> Path:
    return constants.guardian_home() / "ruleset.json"


def _lock_path() -> Path:
    return constants.guardian_home() / "ruleset.lock"


def _serialize(rs: Ruleset) -> Dict[str, Any]:
    return {
        "synced_at": rs.synced_at,
        "injection": [
            {k: v for k, v in rule.items() if k != "pattern"} for rule in rs.injection
        ],
        "escalation": rs.escalation,
        "dependency": rs.dependency,
        "fileaccess": rs.fileaccess,
        "skill": rs.skill,
    }


def _deserialize(data: Dict[str, Any]) -> Ruleset:
    rs = Ruleset(synced_at=float(data.get("synced_at", 0.0)))
    for rule in data.get("injection", []):
        src = rule.get("pattern_src")
        if not src:
            continue
        try:
            compiled = re.compile(src, re.IGNORECASE)
        except re.error:
            continue
        rs.injection.append({**rule, "pattern": compiled})
    rs.escalation = list(data.get("escalation", []))
    rs.dependency = dict(data.get("dependency", {}))
    rs.fileaccess = list(data.get("fileaccess", []))
    rs.skill = list(data.get("skill", []))
    return rs


def _write_cache(rs: Ruleset) -> None:
    try:
        home = constants.guardian_home()
        home.mkdir(parents=True, exist_ok=True)
        tmp = _cache_path().with_suffix(".json.tmp")
        tmp.write_text(json.dumps(_serialize(rs)), encoding="utf-8")
        tmp.replace(_cache_path())
    except Exception as exc:  # pragma: no cover - fail open
        logger.debug("guardian: ruleset cache write failed: %s", exc)


def _read_cache() -> Optional[Ruleset]:
    path = _cache_path()
    if not path.exists():
        return None
    try:
        return _deserialize(json.loads(path.read_text(encoding="utf-8")))
    except Exception as exc:  # pragma: no cover - fail open
        logger.debug("guardian: ruleset cache read failed: %s", exc)
        return None


# ---------------------------------------------------------------------------
# Refresh
# ---------------------------------------------------------------------------

_memory_lock = threading.Lock()
_memory_cache: Optional[Ruleset] = None
_refreshing = False


def refresh(config: Optional[GuardianConfig] = None, client: Optional[DkgClient] = None) -> Ruleset:
    """Query the node, rebuild the ruleset, and persist it. Fail-open.

    Merges curated threats (VM view) with the node's local graph (SWM view).
    On total failure, returns the last-good cache or an empty ruleset — never
    raises.
    """
    global _memory_cache
    config = config or load_guardian_config()
    client = client or DkgClient(url=config.dkg_url)
    rows: List[Any] = []
    got_any = False
    # Public curated graph first (source of truth), then the community pool.
    tiers = (
        (constants.VIEW_VERIFIABLE_MEMORY, "public"),
        (constants.VIEW_SHARED_WORKING_MEMORY, "community"),
    )
    for view, tier in tiers:
        try:
            view_rows = client.query(_THREATS_SPARQL, config.context_graph_id, view=view)
            if view_rows:
                got_any = True
                rows.extend((row, tier) for row in view_rows)
        except Exception as exc:  # pragma: no cover - fail open
            logger.debug("guardian: ruleset query (%s) failed: %s", view, exc)
    if not got_any and not rows:
        # Nothing came back — keep the last-good ruleset instead of emptying.
        existing = _memory_cache or _read_cache()
        if existing is not None:
            existing.synced_at = time.time()
            _write_cache(existing)
            with _memory_lock:
                _memory_cache = existing
            return existing
    rs = build_from_rows(rows)
    _write_cache(rs)
    with _memory_lock:
        _memory_cache = rs
    return rs


def _background_refresh(config: GuardianConfig) -> None:
    global _refreshing
    # Cross-process single-refresher guard via an exclusive lock file.
    lock_fh = None
    try:
        import fcntl

        home = constants.guardian_home()
        home.mkdir(parents=True, exist_ok=True)
        lock_fh = open(_lock_path(), "w")
        try:
            fcntl.flock(lock_fh.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        except (OSError, BlockingIOError):
            lock_fh.close()
            _refreshing = False  # release the in-process guard for the next attempt
            return  # another process is already refreshing
    except Exception:
        lock_fh = None  # platform without fcntl — fall back to in-process guard only
    try:
        refresh(config)
    except Exception as exc:  # pragma: no cover - fail open
        logger.debug("guardian: background refresh failed: %s", exc)
    finally:
        _refreshing = False
        if lock_fh is not None:
            try:
                lock_fh.close()
            except Exception:
                pass


def get(config: Optional[GuardianConfig] = None) -> Ruleset:
    """Return the cached ruleset, lazily refreshing in the background if stale.

    Never blocks on the network: a stale cache is returned immediately while a
    single background thread refreshes it for the next call.
    """
    global _memory_cache, _refreshing
    config = config or load_guardian_config()
    with _memory_lock:
        cached = _memory_cache
    if cached is None:
        cached = _read_cache() or Ruleset()
        with _memory_lock:
            _memory_cache = cached
    age = time.time() - cached.synced_at
    if age > max(1, config.sync_interval) and not _refreshing:
        _refreshing = True
        try:
            threading.Thread(
                target=_background_refresh, args=(config,), name="guardian-ruleset", daemon=True
            ).start()
        except Exception:  # pragma: no cover
            _refreshing = False
    return cached

File: integrations/openclaw/src/ruleset.ts
/**
 * Graph-synced rule cache.
 *
 * The ruleset is the ONLY source of detection truth (mirrors Python
 * `ruleset.py`):
 *
 *   - `verifiable-memory` (the curated public threat graph) → rules tagged
 *     `source: "public"`. The source of truth: matches are CONFIRMED and, in
 *     block mode, blockable.
 * Each tier pulls both legacy Guardian threats and Defender signal entities, normalized into a
 * `Ruleset` and cached to a JSON file (source tags included) under the
 * OpenClaw state dir with a TTL. On an empty graph the ruleset is empty and
 * the matcher detects nothing — by design.
 *
 * Fail-open: if the node is unreachable we keep serving the last cached (or
 * empty) ruleset and try again after the TTL.
 */
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { DkgClient, DkgError, DkgView } from "./dkgClient.js";
import {
  DependencyRule,
  EscalationRule,
  FileAccessRule,
  IocRule,
  InjectionRule,
  RuleSource,
  Ruleset,
  SkillRule,
  emptyRuleset,
} from "./detection.js";
import {
  BLACKBOX_ARG_SHAPE_PRED,
  BLACKBOX_CATEGORY_PRED,
  BLACKBOX_DANGER_SHAPE_PRED,
  BLACKBOX_IDENTIFIER_PRED,
  BLACKBOX_KIND_PRED,
  BLACKBOX_OWASP_CATEGORY_PRED,
  BLACKBOX_PACKAGE_ECOSYSTEM_PRED,
  BLACKBOX_PACKAGE_NAME_PRED,
  BLACKBOX_PACKAGE_VERSION_PRED,
  BLACKBOX_PATTERN_PRED,
  BLACKBOX_SEVERITY_PRED,
  BLACKBOX_SKILL_NAME_PRED,
  BLACKBOX_SKILL_VERSION_PRED,
  BLACKBOX_TOOL_NAME_PRED,
  SCHEMA_DESCRIPTION,
  SCHEMA_NAME,
  RDF_TYPE,
  normalizeSeverity,
} from "./quads.js";

const SCHEMA_ADVISORY = "http://schema.org/identifier";
const DEFENDER = "urn:defender:";
const DEFENDER_P = "urn:defender:p:";

/**
 * SPARQL that pulls legacy threats and the current Defender signal ontology.
 */
function rulesetQuery(): string {
  return `
SELECT ?s ?p ?o WHERE {
  {
    ?s <${BLACKBOX_IDENTIFIER_PRED}> ?identifier .
  }
  UNION
  {
    ?s a ?signalType .
    VALUES ?signalType {
      <${DEFENDER}DependencySignal> <${DEFENDER}InjectionSignal>
      <${DEFENDER}SkillSignal> <${DEFENDER}IocSignal>
    }
  }
  ?s ?p ?o .
}`.trim();
}

/** Agent Blackbox is VM-only until community support ships. */
const TIERS: ReadonlyArray<readonly [DkgView, RuleSource]> = [
  ["verifiable-memory", "public"],
];

export function resolveStateDir(explicit?: string): string {
  if (explicit) return explicit;
  const base =
    process.env.OPENCLAW_STATE_DIR ||
    process.env.OPENCLAW_HOME ||
    join(homedir(), ".openclaw");
  return join(base, "blackbox");
}

/** Extract SPARQL SELECT bindings from the node's response envelope, tolerantly.
 * The v10 node nests them under ``result.bindings`` (singular); older/other
 * shapes use ``results.bindings`` or a top-level ``bindings``. */
function extractBindings(resp: unknown): Array<Record<string, unknown>> {
  const r = resp as Record<string, unknown> | null;
  const container = (r?.result ?? r?.results ?? r) as Record<string, unknown> | undefined;
  const bindings =
    (container as { bindings?: unknown })?.bindings ?? (r as { bindings?: unknown })?.bindings;
  return Array.isArray(bindings) ? (bindings as Array<Record<string, unknown>>) : [];
}

/** Unwrap a binding value: ``{value}`` object, a quoted N-Triples literal
 * (``"x"``, ``"x"^^<type>``, ``"x"@lang``), or a bare IRI/string. */
function bindingValue(v: unknown): string | undefined {
  if (v == null) return undefined;
  if (typeof v === "object" && "value" in (v as Record<string, unknown>)) {
    return String((v as { value: unknown }).value);
  }
  const s = String(v);
  const m = s.match(/^"([\s\S]*)"(?:\^\^.*|@.*)?$/);
  return m ? m[1] : s;
}

interface ThreatAccum {
  subject?: string;
  rdfType?: string;
  identifier?: string;
  severity?: string;
  name?: string;
  description?: string;
  pattern?: string;
  owaspCategory?: string;
  toolName?: string;
  argShape?: string;
  packageName?: string;
  packageVersion?: string;
  packageEcosystem?: string;
  advisoryId?: string;
  // dependency threat kind (malware | vulnerability)
  kind?: string;
  // fileaccess
  category?: string;
  // skill
  skillName?: string;
  skillVersion?: string;
  dangerShape?: string;
  // IOC
  iocValue?: string;
}

/** Accumulate the s/p/o rows of one tier's response into per-subject threats. */
function collectThreats(resp: unknown): ThreatAccum[] {
  const bySubject = new Map<string, ThreatAccum>();
  for (const b of extractBindings(resp)) {
    const s = bindingValue(b.s);
    const p = bindingValue(b.p);
    const o = bindingValue(b.o);
    if (!s || !p || o === undefined) continue;
    const acc = bySubject.get(s) ?? {};
    acc.subject = s;
    switch (p) {
      case RDF_TYPE: acc.rdfType = o; break;
      case BLACKBOX_IDENTIFIER_PRED: acc.identifier = o; break;
      case BLACKBOX_SEVERITY_PRED:
      case `${DEFENDER_P}severity`: acc.severity = o; break;
      case SCHEMA_NAME: acc.name = o; break;
      case SCHEMA_DESCRIPTION: acc.description = o; break;
      case BLACKBOX_PATTERN_PRED:
      case `${DEFENDER_P}pattern`: acc.pattern = o; break;
      case BLACKBOX_OWASP_CATEGORY_PRED: acc.owaspCategory = o; break;
      case BLACKBOX_TOOL_NAME_PRED: acc.toolName = o; break;
      case BLACKBOX_ARG_SHAPE_PRED: acc.argShape = o; break;
      case BLACKBOX_PACKAGE_NAME_PRED:
      case `${DEFENDER_P}package`: acc.packageName = o; break;
      case BLACKBOX_PACKAGE_VERSION_PRED:
      case `${DEFENDER_P}version`: acc.packageVersion = o; break;
      case BLACKBOX_PACKAGE_ECOSYSTEM_PRED:
      case `${DEFENDER_P}ecosystem`: acc.packageEcosystem = o; break;
      case BLACKBOX_KIND_PRED:
      case `${DEFENDER_P}kind`: acc.kind = o; break;
      case SCHEMA_ADVISORY:
      case `${DEFENDER_P}advisoryId`: acc.advisoryId = o; break;
      case BLACKBOX_CATEGORY_PRED:
      case `${DEFENDER_P}iocType`: acc.category = o; break;
      case `${DEFENDER_P}value`: acc.iocValue = o; break;
      case BLACKBOX_SKILL_NAME_PRED: acc.skillName = o; break;
      case BLACKBOX_SKILL_VERSION_PRED: acc.skillVersion = o; break;
      case BLACKBOX_DANGER_SHAPE_PRED: acc.dangerShape = o; break;
      default: break;
    }
    bySubject.set(s, acc);
  }
  return [...bySubject.values()];
}

type MappedRule =
  | { category: "injection"; key: string; rule: InjectionRule }
  | { category: "escalation"; key: string; rule: EscalationRule }
  | { category: "dependency"; key: string; rule: DependencyRule }
  | { category: "fileaccess"; key: string; rule: FileAccessRule }
  | { category: "skill"; key: string; rule: SkillRule }
  | { category: "ioc"; key: string; rule: IocRule };

const QUOTED_SKILL_NAME_RE = /^["'`]([^"'`]+)["'`]/;
const SKILL_PACKAGE_NAME_RE = /^@?[a-z0-9][a-z0-9._-]*(?:\/[a-z0-9][a-z0-9._-]*)?$/i;
const TITLED_SKILL_NAME_RE = /^(@?[a-z0-9][a-z0-9._-]*(?:\/[a-z0-9][a-z0-9._-]*)?)\s+(?:\(|bcc\b)/i;

/** Recover only explicit package-shaped names from legacy display titles. */
export function skillNameFromTitle(title: string): string {
  const value = (title || "").trim();
  return QUOTED_SKILL_NAME_RE.exec(value)?.[1]?.trim()
    ?? TITLED_SKILL_NAME_RE.exec(value)?.[1]?.trim()
    ?? "";
}

/** Map one accumulated threat to `(category, key, rule)` or null. Port of Python `_row_to_rule`. */
function accumToRule(acc: ThreatAccum, source: RuleSource): MappedRule | null {
  let identifier = acc.identifier;
  const suffix = acc.subject?.split(":").pop() ?? "";
  if (!identifier && acc.rdfType === `${DEFENDER}DependencySignal`) {
    const eco = (acc.packageEcosystem || "").toLowerCase();
    const pkg = (acc.packageName || "").toLowerCase();
    const ver = acc.packageVersion || "";
    if (eco && pkg && ver) identifier = `dep:${eco}:${pkg}@${ver}`;
  } else if (!identifier && acc.rdfType === `${DEFENDER}InjectionSignal` && suffix) {
    identifier = `injection:${suffix}`;
  } else if (!identifier && acc.rdfType === `${DEFENDER}SkillSignal` && suffix) {
    identifier = `skill:${suffix}`;
  } else if (!identifier && acc.rdfType === `${DEFENDER}IocSignal` && acc.category && acc.iocValue) {
    identifier = `ioc:${acc.category.trim().toLowerCase()}:${acc.iocValue}`;
  }
  if (!identifier) return null;
  const severity = normalizeSeverity(acc.severity, "high");
  const name = acc.name || acc.description || identifier;
  if (identifier.startsWith("injection:")) {
    if (!acc.pattern) return null;
    return {
      category: "injection",
      key: identifier,
      rule: {
        identifier,
        pattern: acc.pattern,
        severity,
        name,
        owaspCategory: acc.owaspCategory,
        source,
      },
    };
  }
  if (identifier.startsWith("escalation:")) {
    if (!acc.toolName || !acc.argShape) return null;
    return {
      category: "escalation",
      key: identifier,
      rule: { identifier, toolName: acc.toolName, argShape: acc.argShape, severity, name, source },
    };
  }
  if (identifier.startsWith("dep:")) {
    let eco = (acc.packageEcosystem || "").toLowerCase();
    let pkg = (acc.packageName || "").toLowerCase();
    let ver = acc.packageVersion || "";
    if (!(eco && pkg && ver)) {
      // Fall back to parsing the identifier: dep:{eco}:{name}@{version}.
      // Mirrors Python: rest.split(":", 1) then tail.rsplit("@", 1).
      const rest = identifier.slice("dep:".length);
      const ci = rest.indexOf(":");
      if (ci < 0) return null;
      const tail = rest.slice(ci + 1);
      const at = tail.lastIndexOf("@");
      if (at < 0) return null;
      eco = rest.slice(0, ci).toLowerCase();
      pkg = tail.slice(0, at).toLowerCase();
      ver = tail.slice(at + 1);
    }
    return {
      category: "dependency",
      key: `${eco}:${pkg}@${ver}`,
      rule: { identifier, severity, advisoryId: acc.advisoryId, kind: acc.kind || undefined, name, source },
    };
  }
  if (identifier.startsWith("fileaccess:")) {
    // Prefer explicit predicates; fall back to parsing fileaccess:{tool}:{category}.
    let toolName = acc.toolName;
    let category = acc.category;
    if (!toolName || !category) {
      const parts = identifier.split(":");
      // ["fileaccess", tool, category] — mirror Python identifier.split(":", 2).
      if (parts.length >= 3) {
        toolName = parts[1];
        category = parts.slice(2).join(":");
      }
    }
    if (!toolName || !category) return null;
    return {
      category: "fileaccess",
      key: identifier,
      rule: {
        identifier,
        toolName: toolName.trim().toLowerCase(),
        category: category.trim().toLowerCase(),
        severity,
        name,
        source,
      },
    };
  }
  if (identifier.startsWith("skill:")) {
    // Skill rules are built unconditionally (fields optional), like Python.
    return {
      category: "skill",
      key: identifier,
      rule: {
        identifier,
        skillName: acc.skillName && SKILL_PACKAGE_NAME_RE.test(acc.skillName)
          ? acc.skillName
          : skillNameFromTitle(name),
        skillVersion: acc.skillVersion ?? "",
        dangerShape: acc.dangerShape ?? "",
        severity,
        name,
        source,
      },
    };
  }
  if (identifier.startsWith("ioc:")) {
    const parts = identifier.split(":", 3);
    const iocType = (acc.category || parts[1] || "").trim().toLowerCase();
    const value = acc.iocValue || identifier.slice(`ioc:${parts[1] ?? ""}:`.length);
    if (!iocType || !value) return null;
    return {
      category: "ioc",
      key: identifier,
      rule: {
        identifier,
        severity,
        name,
        iocType,
        value,
        kind: acc.kind || undefined,
        source,
      },
    };
  }
  return null;
}

/**
 * Build one merged `Ruleset` from per-tier threat accumulations, in tier
 * order (public first). Precedence is identifier-first-wins in every category
 * INCLUDING the dependency map, so with public processed first a community
 * row can never shadow (or escalate/downgrade) a curated public rule. Port of
 * Python `build_from_rows`.
 */
function buildRuleset(tiers: ReadonlyArray<readonly [ThreatAccum[], RuleSource]>): Ruleset {
  const ruleset = emptyRuleset();
  ruleset.fetchedAt = Date.now();
  const seen: Record<MappedRule["category"], Set<string>> = {
    injection: new Set(),
    escalation: new Set(),
    dependency: new Set(),
    fileaccess: new Set(),
    skill: new Set(),
    ioc: new Set(),
  };
  for (const [accums, source] of tiers) {
    for (const acc of accums) {
      const mapped = accumToRule(acc, source);
      if (!mapped || seen[mapped.category].has(mapped.key)) continue;
      seen[mapped.category].add(mapped.key);
      switch (mapped.category) {
        case "injection": ruleset.injection.push(mapped.rule); break;
        case "escalation": ruleset.escalation.push(mapped.rule); break;
        case "dependency": ruleset.dependency[mapped.key] = mapped.rule; break;
        case "fileaccess": ruleset.fileaccess.push(mapped.rule); break;
        case "skill": ruleset.skill.push(mapped.rule); break;
        case "ioc": ruleset.ioc[mapped.key] = mapped.rule; break;
      }
    }
  }
  return ruleset;
}

export interface RulesetCacheOptions {
  client: DkgClient;
  contextGraphId: string;
  stateDir?: string;
  /** TTL seconds; below this a get() serves the cache without a refresh. */
  ttlSeconds?: number;
}

/**
 * Lazy, TTL'd ruleset cache. `get()` returns the in-memory ruleset immediately
 * and triggers a background refresh when stale; the refresh is fail-open.
 */
export class RulesetCache {
  private readonly client: DkgClient;
  private readonly contextGraphId: string;
  private readonly cachePath: string;
  private readonly ttlMs: number;
  private ruleset: Ruleset;
  private refreshing = false;

  constructor(opts: RulesetCacheOptions) {
    this.client = opts.client;
    this.contextGraphId = opts.contextGraphId;
    this.ttlMs = (opts.ttlSeconds ?? 300) * 1000;
    this.cachePath = join(resolveStateDir(opts.stateDir), "ruleset.json");
    this.ruleset = this.loadFromDisk() ?? emptyRuleset();
  }

  private loadFromDisk(): Ruleset | null {
    try {
      const parsed = JSON.parse(readFileSync(this.cachePath, "utf8")) as Partial<Ruleset>;
      if (parsed && Array.isArray(parsed.injection)) {
        const publicOnly = <T extends { source?: RuleSource }>(rows: T[] | undefined): T[] =>
          (rows ?? []).filter((row) => row.source !== "community");
        const dependencies = Object.fromEntries(
          Object.entries(parsed.dependency ?? {}).filter(
            ([, rule]) => rule?.source !== "community",
          ),
        );
        // Backfill arrays absent from caches written before fileaccess/skill
        // support and drop SWM rows written by pre-VM-only releases.
        return {
          ...emptyRuleset(),
          ...parsed,
          injection: publicOnly(parsed.injection),
          escalation: publicOnly(parsed.escalation),
          dependency: dependencies,
          fileaccess: publicOnly(parsed.fileaccess),
          skill: publicOnly(parsed.skill),
          ioc: Object.fromEntries(
            Object.entries(parsed.ioc ?? {}).filter(([, rule]) => rule?.source !== "community"),
          ),
        } as Ruleset;
      }
    } catch {
      /* no cache yet */
    }
    return null;
  }

  private saveToDisk(rs: Ruleset): void {
    try {
      mkdirSync(dirname(this.cachePath), { recursive: true });
      writeFileSync(this.cachePath, JSON.stringify(rs), "utf8");
    } catch {
      /* best effort */
    }
  }

  private isStale(): boolean {
    return Date.now() - this.ruleset.fetchedAt > this.ttlMs;
  }

  /**
   * Force a synchronous sync from the node. Fail-open (returns current on
   * error). One query per sync reads verifiable-memory. If nothing comes back,
   * the last-good public ruleset keeps serving. Mirrors Python `ruleset.refresh`.
   */
  async sync(): Promise<Ruleset> {
    if (this.refreshing) return this.ruleset;
    this.refreshing = true;
    try {
      const tierAccums: Array<readonly [ThreatAccum[], RuleSource]> = [];
      let gotAny = false;
      for (const [view, source] of TIERS) {
        try {
          const resp = await this.client.query(rulesetQuery(), this.contextGraphId, view);
          const accums = collectThreats(resp);
          if (accums.length > 0) gotAny = true;
          tierAccums.push([accums, source]);
        } catch (err) {
          if (!(err instanceof DkgError)) throw err;
          // this view unreachable/failed — skip it, keep the other tier.
        }
      }
      if (!gotAny) {
        // Nothing came back — keep serving the last ruleset, bump fetchedAt so
        // we don't hammer a down node every tool call.
        this.ruleset = { ...this.ruleset, fetchedAt: Date.now() };
      } else {
        const next = buildRuleset(tierAccums);
        this.ruleset = next;
        this.saveToDisk(next);
      }
    } finally {
      this.refreshing = false;
    }
    return this.ruleset;
  }

  /** Cached ruleset; kicks off a fire-and-forget refresh when stale. */
  get(): Ruleset {
    if (this.isStale() && !this.refreshing) {
      void this.sync();
    }
    return this.ruleset;
  }

  counts(): {
    injection: number;
    escalation: number;
    dependency: number;
    fileaccess: number;
    skill: number;
    ioc: number;
  } {
    return {
      injection: this.ruleset.injection.length,
      escalation: this.ruleset.escalation.length,
      dependency: Object.keys(this.ruleset.dependency).length,
      fileaccess: this.ruleset.fileaccess.length,
      skill: this.ruleset.skill.length,
      ioc: Object.keys(this.ruleset.ioc).length,
    };
  }
}
